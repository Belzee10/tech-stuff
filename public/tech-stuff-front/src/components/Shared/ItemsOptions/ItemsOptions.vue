<template>
  <v-row>
    <v-col cols="8">
      <v-row>
        <v-col cols="6">
          <v-text-field
            dense
            clearable
            label="Search"
            @input="value => $emit('search', value)"
            @click:clear="() => $emit('clear-search')"
          >
            <v-icon slot="prepend">mdi-magnify</v-icon>
          </v-text-field>
        </v-col>
      </v-row>
    </v-col>
    <v-col cols="4">
      <v-row>
        <v-col cols="6">
          <v-select
            flat
            dense
            hide-details
            outlined
            :items="getSortOptions()"
            :value="sortValue"
            label="Sort"
            :menu-props="{ bottom: true, offsetY: true }"
            @change="value => $emit('sort-by-price', value)"
          >
            <template v-slot:item="{ item }">
              <v-icon small class="mr-1">mdi-{{ item.icon }}</v-icon>
              <span>{{ item.text }}</span>
            </template>
            <template v-slot:append>
              <v-icon small>mdi-{{ getIcon }}</v-icon>
            </template>
          </v-select>
        </v-col>
        <v-col cols="6">
          <v-btn-toggle
            color="primary"
            :value="viewValue"
            @change="value => $emit('change-view', value)"
          >
            <v-btn value="cards" class="cards" title="Grid view">
              <v-icon small>mdi-view-grid</v-icon>
            </v-btn>
            <v-btn value="items" class="items" title="Items view">
              <v-icon small>mdi-format-list-bulleted-square</v-icon>
            </v-btn>
          </v-btn-toggle>
        </v-col>
      </v-row>
    </v-col>
  </v-row>
</template>

<script>
export default {
  name: 'ItemsOptions',
  props: {
    viewValue: {
      type: String,
      default: ''
    },
    sortValue: {
      type: String,
      default: ''
    }
  },
  computed: {
    getIcon() {
      const item = this.getSortOptions().find(
        item => item.value === this.sortValue
      );
      return item ? item.icon : '';
    }
  },
  methods: {
    getSortOptions() {
      return [
        {
          text: 'Asc',
          value: 'ASC',
          icon: 'sort-ascending'
        },
        {
          text: 'Desc',
          value: 'DESC',
          icon: 'sort-descending'
        }
      ];
    }
  }
};
</script>

<style lang="scss" scoped>
.v-select,
.v-item-group {
  border-radius: 2px !important;
}
.v-btn {
  height: 40px !important;
}
</style>
