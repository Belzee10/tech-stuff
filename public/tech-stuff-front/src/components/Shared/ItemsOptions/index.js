export { default } from './ItemsOptions.vue';
