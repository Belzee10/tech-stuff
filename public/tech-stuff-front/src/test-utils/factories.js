import { mount, createLocalVue } from '@vue/test-utils';
import mergeWith from 'lodash.mergewith';
import Vuetify from 'vuetify';
import Vuex from 'vuex';

const localVue = createLocalVue();
localVue.use(Vuex);

let vuetify;

beforeEach(() => {
  vuetify = new Vuetify();
});

const customizer = (_, srcValue) => {
  if (Array.isArray(srcValue)) {
    return srcValue;
  }
  if (srcValue instanceof Object && Object.keys(srcValue).length === 0) {
    return srcValue;
  }
};

/**
 * create component wrapper
 * @param {*} Component
 * @param {Object} overrides
 */
const createWrapper = (Component, overrides) => {
  const defaultMountingOptions = {
    localVue,
    vuetify
  };
  return mount(
    Component,
    mergeWith(defaultMountingOptions, overrides, customizer)
  );
};

/**
 * create Vuex store
 * @param {Object} overrides
 */
const createStore = overrides => {
  const defaultStoreConfig = {};
  return new Vuex.Store(mergeWith(defaultStoreConfig, overrides, customizer));
};

export { createWrapper, createStore };
