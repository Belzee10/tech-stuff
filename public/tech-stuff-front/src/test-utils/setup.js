import Vue from 'vue';
import Vuetify from 'vuetify';

Vue.config.productionTip = false;

Vue.use(Vuetify);

/**
 * silent "console.error"
 */
global.console = {
  error: jest.fn(),
  log: console.log,
  warn: console.warn,
  info: console.info,
  debug: console.debug
};
