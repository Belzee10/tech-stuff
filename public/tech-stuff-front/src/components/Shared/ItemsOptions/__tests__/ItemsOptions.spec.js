import { createWrapper } from '@/test-utils/factories';
import ItemsOptions from '../ItemsOptions.vue';

describe('ItemsOptions.vue', () => {
  test('should emit a "change-view" event', () => {
    const props = {
      viewValue: 'cards'
    };
    const wrapper = createWrapper(ItemsOptions, {
      propsData: props
    });
    wrapper.find('.v-item-group button.items').trigger('click');
    expect(wrapper.emitted('change-view')[0]).toEqual(['items']);
  });

  test('should emit a "sort-by-price" event', () => {
    const props = {
      sortValue: 'ASC'
    };
    const wrapper = createWrapper(ItemsOptions, {
      propsData: props
    });
    wrapper.find('.v-select').vm.$emit('change', 'DESC');
    expect(wrapper.emitted('sort-by-price')[0]).toEqual(['DESC']);
  });

  test('should emit a "search" event', () => {
    const value = 'q';
    const wrapper = createWrapper(ItemsOptions);
    wrapper.find('.v-text-field input').setValue(value);
    expect(wrapper.emitted('search')[0]).toEqual([value]);
  });

  test('should render correctly', () => {
    const props = {
      sortValue: 'ASC',
      viewValue: 'cards'
    };
    const wrapper = createWrapper(ItemsOptions, {
      propsData: props
    });
    expect(wrapper.element).toMatchSnapshot();
  });
});
