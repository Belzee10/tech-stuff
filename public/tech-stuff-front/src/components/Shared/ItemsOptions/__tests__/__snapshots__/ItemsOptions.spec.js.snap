// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemsOptions.vue should render correctly 1`] = `
<div
  class="row"
>
  <div
    class="col col-8"
  >
    <div
      class="row"
    >
      <div
        class="col col-6"
      >
        <div
          class="v-input v-input--dense theme--light v-text-field"
        >
          <div
            class="v-input__prepend-outer"
          >
            <i
              aria-hidden="true"
              class="v-icon notranslate mdi mdi-magnify theme--light"
            />
          </div>
          <div
            class="v-input__control"
          >
            <div
              class="v-input__slot"
            >
              <div
                class="v-text-field__slot"
              >
                <label
                  class="v-label theme--light"
                  for="input-78"
                  style="left: 0px; position: absolute;"
                >
                  Search
                </label>
                <input
                  id="input-78"
                  type="text"
                />
              </div>
              <div
                class="v-input__append-inner"
              >
                <div
                  class="v-input__icon v-input__icon--"
                >
                  <i
                    class="v-icon notranslate v-icon--link material-icons theme--light"
                    role="button"
                  />
                </div>
              </div>
            </div>
            <div
              class="v-text-field__details"
            >
              <div
                class="v-messages theme--light"
              >
                <span
                  class="v-messages__wrapper"
                  name="message-transition"
                  tag="div"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
   
  <div
    class="col col-4"
  >
    <div
      class="row"
    >
      <div
        class="col col-6"
      >
        <div
          class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty v-input--dense theme--light v-text-field v-text-field--solo-flat v-text-field--enclosed v-text-field--outlined v-select"
        >
          <div
            class="v-input__control"
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-owns="list-83"
              class="v-input__slot"
              role="button"
            >
              <fieldset
                aria-hidden="true"
              >
                <legend
                  style="width: 6px;"
                >
                  <span>
                    â€‹
                  </span>
                </legend>
              </fieldset>
              <div
                class="v-select__slot"
              >
                <label
                  class="v-label v-label--active theme--light"
                  for="input-83"
                  style="left: 0px; position: absolute;"
                >
                  Sort
                </label>
                <div
                  class="v-select__selections"
                >
                  <div
                    class="v-select__selection v-select__selection--comma"
                  >
                    Asc
                  </div>
                  <input
                    aria-readonly="true"
                    id="input-83"
                    readonly="readonly"
                    type="text"
                  />
                </div>
                <div
                  class="v-input__append-inner"
                >
                  <i
                    aria-hidden="true"
                    class="v-icon notranslate mdi mdi-sort-ascending theme--light"
                    style="font-size: 16px;"
                  />
                </div>
              </div>
              <div
                class="v-menu"
              >
                <div
                  class="v-menu__content theme--light "
                  style="max-height: 304px; min-width: 0px; max-width: auto; top: 12px; left: 0px; transform-origin: top left; z-index: 0; display: none;"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
       
      <div
        class="col col-6"
      >
        <div
          class="v-item-group theme--light v-btn-toggle primary--text"
        >
          <button
            class="cards v-btn v-item--active v-btn--active v-btn--contained theme--light v-size--default"
            title="Grid view"
            type="button"
            value="cards"
          >
            <span
              class="v-btn__content"
            >
              <i
                aria-hidden="true"
                class="v-icon notranslate mdi mdi-view-grid theme--light"
                style="font-size: 16px;"
              />
            </span>
          </button>
           
          <button
            class="items v-btn v-btn--contained theme--light v-size--default"
            title="Items view"
            type="button"
            value="items"
          >
            <span
              class="v-btn__content"
            >
              <i
                aria-hidden="true"
                class="v-icon notranslate mdi mdi-format-list-bulleted-square theme--light"
                style="font-size: 16px;"
              />
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
